File: card-core/src/main/java/com/jean202/cardmizer/core/domain/RecommendationContext.java
package com.jean202.cardmizer.core.domain;

import com.jean202.cardmizer.common.Money;
import java.util.LinkedHashSet;
import java.util.Locale;
import java.util.Objects;
import java.util.Set;

public record RecommendationContext(
        SpendingPeriod spendingPeriod,
        Money amount,
        String merchantName,
        String merchantCategory,
        Set<String> paymentTags
) {
    public RecommendationContext(SpendingPeriod spendingPeriod, Money amount, String merchantName, String merchantCategory) {
        this(spendingPeriod, amount, merchantName, merchantCategory, Set.of());
    }

    public RecommendationContext {
        Objects.requireNonNull(spendingPeriod, "spendingPeriod must not be null");
        Objects.requireNonNull(amount, "amount must not be null");
        if (merchantName == null || merchantName.isBlank()) {
            throw new IllegalArgumentException("Merchant name must not be blank");
        }
        merchantName = merchantName.trim();
        merchantCategory = normalizeValue(merchantCategory == null || merchantCategory.isBlank()
                ? "UNCATEGORIZED"
                : merchantCategory);
        paymentTags = normalizeValues(paymentTags);
    }

    private static String normalizeValue(String value) {
        return value.trim().toUpperCase(Locale.ROOT);
    }

    private static Set<String> normalizeValues(Set<String> rawTags) {
        if (rawTags == null || rawTags.isEmpty()) {
            return Set.of();
        }
        Set<String> normalizedTags = new LinkedHashSet<>();
        for (String rawTag : rawTags) {
            if (rawTag != null && !rawTag.isBlank()) {
                normalizedTags.add(normalizeValue(rawTag));
            }
        }
        return Set.copyOf(normalizedTags);
    }
}

File: card-core/src/main/kotlin/com/jean202/cardmizer/core/domain/RecommendationContext.kt
package com.jean202.cardmizer.core.domain

import com.jean202.cardmizer.common.Money
import java.util.Locale

class RecommendationContext(
    val spendingPeriod: SpendingPeriod,
    val amount: Money,
    merchantName: String,
    merchantCategory: String? = null,
    paymentTags: Set<String> = emptySet(),
) {
    val merchantName: String
    val merchantCategory: String
    val paymentTags: Set<String>

    init {
        require(merchantName.isNotBlank()) { "Merchant name must not be blank" }
        this.merchantName = merchantName.trim()
        this.merchantCategory =
            if (merchantCategory.isNullOrBlank()) "UNCATEGORIZED"
            else normalizeValue(merchantCategory)
        this.paymentTags = normalizeValues(paymentTags)
    }

    companion object {
        private fun normalizeValue(value: String): String =
            value.trim().uppercase(Locale.ROOT)

        private fun normalizeValues(rawTags: Set<String>): Set<String> =
            rawTags.filter { it.isNotBlank() }.map { normalizeValue(it) }.toSet()
    }
}
